clog/clog.go

```diff
@@ -0,0 +1,196 @@
+	newCmap := newValues()
+	if cmap != nil {
+		cmap.mu.RLock()
+		for k, v := range cmap.vals {
+			newCmap.vals[k] = v
+		}
+		cmap.mu.RUnlock()
+	}
+	return context.WithValue(parentCtx, clogContextKey, newCmap)
+}
+
+// Same creates new context with parentCtx as parent and
+// same logging data as in logCtx
+func Same(parentCtx, logCtx context.Context) context.Context {
+	cmap, _ := logCtx.Value(clogContextKey).(*values)
+	if cmap == nil {
+		cmap = newValues()
+	}
+	return context.WithValue(parentCtx, clogContextKey, cmap)
+}
+
+func AddManifestID(ctx context.Context, val string) context.Context {
+	return AddVal(ctx, manifestID, val)
+}
+
+func AddOrchSessionID(ctx context.Context, val string) context.Context {
+	return AddVal(ctx, orchSessionID, val)
+}
+
+func AddSessionID(ctx context.Context, val string) context.Context {
+	return AddVal(ctx, sessionID, val)
+}
+
+func AddNonce(ctx context.Context, val uint64) context.Context {
+	return AddVal(ctx, nonce, strconv.FormatUint(val, 10))
+}
+
+func AddSeqNo(ctx context.Context, val uint64) context.Context {
+	return AddVal(ctx, seqNo, strconv.FormatUint(val, 10))
+}
+
+func AddVal(ctx context.Context, key, val string) context.Context {
+	cmap, _ := ctx.Value(clogContextKey).(*values)
+	if cmap == nil {
+		cmap = newValues()
+		ctx = context.WithValue(ctx, clogContextKey, cmap)
+	}
+	cmap.mu.Lock()
+	cmap.vals[key] = val
+	cmap.mu.Unlock()
+	return ctx
+}
+
+func Warningf(ctx context.Context, format string, args ...interface{}) {
+	glog.WarningDepth(1, formatMessage(ctx, false, format, args...))
+}
+
+func Errorf(ctx context.Context, format string, args ...interface{}) {
+	glog.ErrorDepth(1, formatMessage(ctx, false, format, args...))
+}
+
+func Fatalf(ctx context.Context, format string, args ...interface{}) {
+	glog.FatalDepth(1, formatMessage(ctx, false, format, args...))
+}
+
+func Infof(ctx context.Context, format string, args ...interface{}) {
+	infof(ctx, false, format, args...)
+}
+
+// Infofe if last argument is not nil it will be printed as " err=%q"
+func Infofe(ctx context.Context, format string, args ...interface{}) {
+	infof(ctx, true, format, args...)
+}
+
+func infof(ctx context.Context, lastErr bool, format string, args ...interface{}) {
+	glog.InfoDepth(2, formatMessage(ctx, lastErr, format, args...))
+}
+
+// Infof is equivalent to the global Infof function, guarded by the value of v.
+// See the documentation of V for usage.
+func (v Verbose) Infof(ctx context.Context, format string, args ...interface{}) {
+	if v {
+		infof(ctx, false, format, args...)
+	}
+}
+
+func (v Verbose) Infofe(ctx context.Context, format string, args ...interface{}) {
+	if v {
+		infof(ctx, true, format, args...)
+	}
+}
+
+func messageFromContext(ctx context.Context, sb *strings.Builder) {
+	if ctx == nil {
+		return
+	}
+	cmap, _ := ctx.Value(clogContextKey).(*values)
+	if cmap == nil {
+		return
+	}
+	cmap.mu.RLock()
+	for _, key := range stdKeysOrder {
+		if val, ok := cmap.vals[key]; ok {
```

Comment:
>  Is that intended?

It is consequence of `go`'s map being random. I could overcome it by different ways, but I've decided that it's not worth it. And anyway, right now only standard keys are used.

Reply:
If we prepare the `clog` only for standard keys now, then let's maybe print only them? And skip printing the random keys? I think we should make it clear if the user of `clog` is allowed to create custom keys or not. If yes, then be consistent and print them in some order (i.e. alphabetic) if not, then just do not log the custom keys.

Reply:
1. It is allowed to create custom keys.
2. I don't think it's worth it. To do that we will need to write more code and suffer runtime efficiency. We can change it later if we will decide that this is a problem.

Reply:
> 1. It is allowed to create custom keys.

Do we use it anywhere? Why do we need it?

Reply:
I think it will be quite annoying to go through logs where the keys have a random order on every line. Especially when for example going through a stream of the same log type, I normally get used to find the data I want on the same position on the line (cause fields generally have have the same length) and just move around. 

Totally something that we can change in implementation later without breaking compatibility, so not necessary now, but giving my +1 here that we probably want to do some deterministic key order here soon. We could create an auxiliary `keys` slice, sort that, and then print the fields the same way we do for the `stdKeysOrder` above. That's exactly how `logrus` does it ([src](https://github.com/sirupsen/logrus/blob/master/text_formatter.go#L139)) and I'd guess how `json` package does it too for serialization.